src/core/Decoder.ts

```diff
@@ -31,7 +31,39 @@ export type Next<O, I = unknown> =
       err: (msg: string | Annotation) => DecodeResult<O>,
     ) => DecodeResult<O> | Decoder<O>);
 
+export interface ReadonlyDecoder<T> extends Decoder<T> {
+  readonly isReadonly: true;
```

Comment:
This property can be accessed at runtime, to change the behavior of composite decoders like `array` or `object` to not perform a deep copy later.